gtp2ogs.js

```diff
@@ -169,11 +180,160 @@ class Bot {
+                let white_timeleft = Math.max( Math.floor(state.clock.white_time.thinking_time
+                    - white_offset + (state.clock.white_time.periods - 1) * state.time_control.periods), 0);
+
+                this.command("time_settings " + (state.time_control.main_time + (state.time_control.periods - 1) * state.time_control.period_time) + " "
+                    + Math.floor(state.time_control.period_time -
+                        (state.clock.current_player == state.clock.black_player_id
+                            ? (black_timeleft > 0 ? 0 : black_offset) : (white_timeleft > 0 ? 0 : white_offset)
+                        )
+                    )
+                    + " 1");
+                // Since we're faking byoyomi using Canadian, time_left actually does mean the time left to play our 1 stone.
+                //
+                this.command("time_left black " + (black_timeleft > 0 ? black_timeleft + " 0"
+                    : Math.floor(state.time_control.period_time - black_offset) + " 1") );
+                this.command("time_left white " + (white_timeleft > 0 ? white_timeleft + " 0"
+                    : Math.floor(state.time_control.period_time - white_offset) + " 1") );
+            }
+        } else if (state.time_control.system == 'canadian') {
+            // Canadian Byoyomi is the only time controls GTP v2 officially supports.
+            // 
+            let black_timeleft = Math.max( Math.floor(state.clock.black_time.thinking_time - black_offset), 0);
+            let white_timeleft = Math.max( Math.floor(state.clock.white_time.thinking_time - white_offset), 0);
+
+            if (KGSTIME) {
+                this.command("kgs-time_settings canadian " + state.time_control.main_time + " "
+                    + state.time_control.period_time + " " + state.time_control.stones_per_period);
+            } else {
+                this.command("time_settings " + state.time_control.main_time + " "
+                    + state.time_control.period_time + " " + state.time_control.stones_per_period);
+            }
+
+            this.command("time_left black " + (black_timeleft > 0 ? black_timeleft + " 0"
+                : Math.floor(state.clock.black_time.block_time - black_offset) + " " + state.clock.black_time.moves_left));
+            this.command("time_left white " + (white_timeleft > 0 ? white_timeleft + " 0"
+                : Math.floor(state.clock.white_time.block_time - white_offset) + " " + state.clock.white_time.moves_left));
+        } else if (state.time_control.system == 'fischer') {
+            // Not supported by kgs-time_settings and I assume most bots. So lets just handle it like absolute time. A bot may move quicker
+            // than it needs to when low on time, thinking this is total time left for whole game, but since time is added
+            // each move it should be somewhat self-correcting.
```

Comment:
If this is causing the bots to think they only have such low time for the rest of the game, then maybe it would be better to treat it as byo-yomi with one period left of `time_increment` seconds, and main time being the time that's left minus `time_increment` ?